### DISCUSSION TITLE
Import Valhalla as Project Dependency From Source

### DISCUSSION THREAD

**Discussion by LZhaoM** [valhalla/valhalla#4694] [Q&A] [answered]
Hi,

I have successfully built Valhalla from source by using `vcpkg`. Then I try to import libvalhalla as a dependency from source but I don't know where to start. The [documentation](https://valhalla.github.io/valhalla/building/) is very helpful, but I'm still a bit confused after reading it. I am new to C++ and cmake. The following is what I have done:

Project layout:
```
.
├── build/
├── CMakeLists.txt
├── main.cc
├── valhalla/
├── vcpkg/
├── vcpkg_installed/
├── vcpkg-configuration.json
└── vcpkg.json
``` 
CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION 3.14 FATAL_ERROR)

# set the project name and version
project(BWRP LANGUAGES CXX C)

set(PROJECT_BINARY_DIR "${PROJECT_SOURCE_DIR}/build")

# specify the C++ standard
set(CMAKE_CXX_STANDARD 17 CACHE STRING "C++ language version to use (default is 17)")
set(CMAKE_CXX_STANDARD_REQUIRED ON)

include(FindPkgConfig)

add_subdirectory(valhalla)

pkg_check_modules(Valhalla REQUIRED libvalhalla)

add_executable(Run main.cc)

target_link_libraries(Run PUBLIC "libvalhalla")
```
vcpkg.json:
```json
{
  "dependencies": [
    "libspatialite"
  ]
}
```
What I got after running `cmake -B build`:
```
-- No build type specified, defaulting to Release
-- Configuring in release mode
-- Using protoc from /usr/bin/protoc-25.3.0
-- Using pbf headers from /usr/include
-- Using pbf libs from /usr/lib/libprotobuf.so.25.3.0
-- Using pbf-lite
CMake Warning at valhalla/CMakeLists.txt:204 (message):
  Unable to enable gdal support


-- Found SQLite3: /usr/lib/libsqlite3.so
-- Checking for module 'spatialite'
--   Package 'spatialite', required by 'virtual:world', not found
CMake Error at /usr/share/cmake/Modules/FindPkgConfig.cmake:634 (message):
  The following required packages were not found:

   - spatialite

Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindPkgConfig.cmake:862 (_pkg_check_modules_internal)
  valhalla/CMakeLists.txt:231 (pkg_check_modules)


-- Configuring incomplete, errors occurred!
```

I have download `libspatialite` using vcpkg, but I don't know why it is still not found.

Any help is appreciated. Thank you.

**Comment by nilsnolde** [comment, marked as answer]
I wouldn’t recommend including libvalhalla like that. Install it properly to your system first and just remove that add_subdirectory line. That’s the (more) painless route.

**Reply by LZhaoM** [reply]
Thank you so much for the reply. I will definitely try to do that, however, I still have a question: would it be possible to include libvalhalla without installing it system-widely? Or that method is the most convenient one?

**Reply by nilsnolde** [reply]
You can install it where you want if you can convince cmake to go look there. Easiest is probably to set the PKG_CONFIG_PATH env var to …/lib/pkgconfig (where that is when you installed it).

**Reply by LZhaoM** [reply]
I will try that. Thank you so much!
